### PR TITLE
Clean up Expression parameters and the 'analyze' method (part 1)

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Ready for another look, I rewrote queue as a linked list. I'll add `linkedlistof()` typecheckers later, when we fix the typecheck copying problem.

File: python/hail/utils/tests.py [hunk 1]
```diff
@@ -59,27 +59,26 @@ def test_hadoop_methods(self):
...
+        ll = LinkedList(int)
+        self.assertEqual(list(ll), [])
+        if ll:
+            self.fail('empty linked list had an implicit boolean value of True')
 
-        q = q.push(5).push(2)
+        ll2 = ll.push(5).push(2)
 
-        self.assertEqual(list(q), [5, 2])
-        self.assertEqual(len(q), 2)
```
Comment (catoverdrive) [inline python/hail/utils/tests.py]: did you want a `LinkedList.__len__` method? Slightly more expensive unless you store the length in the node, idk if it's useful

Comment (tpoterba) [reply python/hail/utils/tests.py]: don't really need to use it anywhere, so parsimony says no.

Comment (catoverdrive) [reply python/hail/utils/tests.py]: 👍

**Pull request merged**